packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -145,65 +171,426 @@ String getGradleVersionForAndroidPlugin(Directory directory, Logger logger) {
+  }
+  // Check if versions are newer than the max known versions.
+  if (isWithinVersionRange(agpV,
+      min: _maxKnownAndSupportedGradleVersion, max: '100.100')) {
+    // Assume versions we do not know about are valid but log.
+    final bool validGradle =
+        isWithinVersionRange(gradleV, min: '8.0', max: '100.00');
+    logger.printTrace('Newer than known AGP version ($agpV), gradle ($gradleV).'
+        '\n Treating as valid configuration.');
+    return validGradle;
+  }
+
+  // Begin Known Gradle <-> AGP validation.
+  // Max agp here is a made up version to contain all 7.4 changes.
+  if (isWithinVersionRange(agpV, min: '7.4', max: '7.5')) {
+    return isWithinVersionRange(gradleV,
+        min: '7.5', max: _maxKnownAndSupportedGradleVersion);
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '1.2.0', max: '1.3.1')) {
-    return '2.9';
+  if (isWithinVersionRange(agpV,
+      min: '7.3', max: '7.4', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '7.4', max: _maxKnownAndSupportedGradleVersion);
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '1.5.0', max: '1.5.0')) {
-    return '2.2.1';
+  if (isWithinVersionRange(agpV,
+      min: '7.2', max: '7.3', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '7.3.3', max: _maxKnownAndSupportedGradleVersion);
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '2.0.0', max: '2.1.2')) {
-    return '2.13';
+  if (isWithinVersionRange(agpV,
+      min: '7.1', max: '7.2', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '7.2', max: _maxKnownAndSupportedGradleVersion);
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '2.1.3', max: '2.2.3')) {
-    return '2.14.1';
+  if (isWithinVersionRange(agpV,
+      min: '7.0', max: '7.1', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '7.0', max: _maxKnownAndSupportedGradleVersion);
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '2.3.0', max: '2.9.9')) {
-    return '3.3';
+  if (isWithinVersionRange(agpV,
+      min: '4.2.0', max: '7.0', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '6.7.1', max: _maxKnownAndSupportedGradleVersion);
   }
-  if (_isWithinVersionRange(androidPluginVersion, min: '3.0.0', max: '3.0.9')) {
-    return '4.1';
+  if (isWithinVersionRange(agpV,
+      min: '4.1.0', max: '4.2.0', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '6.5', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(agpV,
+      min: '4.0.0', max: '4.1.0', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '6.1.1', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(
+    agpV,
+    min: '3.6.0',
+    max: '3.6.4',
+  )) {
+    return isWithinVersionRange(gradleV,
+        min: '5.6.4', max: _maxKnownAndSupportedGradleVersion);
+  }
```

Comment:
This could benefit from the same loop-through-a-list treatment as in `getGradleVersionFor`.